indexer/tasks/src/multiera/multiera_projected_nft.rs

```diff
@@ -0,0 +1,373 @@
+        .join(
+            JoinType::InnerJoin,
+            ProjectedNftRelation::TransactionOutput.def(),
+        )
+        .join(JoinType::InnerJoin, ProjectedNftRelation::Transaction.def())
+        .filter(inputs_condition)
+        .into_model::<ProjectedNftInputsQueryOutputResult>()
+        .all(db_tx)
+        .await?;
+
+    let mut result: BTreeMap<Vec<u8>, BTreeMap<i64, ProjectedNftInputsQueryOutputResult>> =
+        BTreeMap::new();
+    for nft in projected_nfts {
+        result
+            .entry(nft.tx_hash.clone())
+            .or_default()
+            .entry(nft.output_index as i64)
+            .or_insert(nft);
+    }
+    Ok(result)
+}
+
+#[derive(Debug, Clone)]
+struct ProjectedNftData {
+    pub previous_utxo_tx_hash: Vec<u8>,
+    pub previous_utxo_tx_output_index: Option<i64>,
+    pub address: Vec<u8>,
+    pub plutus_data: Vec<u8>,
+    pub operation: ProjectedNftOperation,
+    pub for_how_long: Option<i64>,
+}
+
+fn extract_operation_and_datum(output: &MultiEraOutput) -> ProjectedNftData {
+    let datum_option = match output.datum() {
+        Some(datum) => DatumOption::from(datum.clone()),
+        None => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
+                previous_utxo_tx_output_index: None,
+                address: vec![],
+                plutus_data: vec![],
+                operation: ProjectedNftOperation::NoDatum,
+                for_how_long: None,
+            };
+        }
+    };
+
+    let datum = match datum_option {
+        DatumOption::Hash(hash) => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
+                previous_utxo_tx_output_index: None,
+                address: vec![],
+                plutus_data: hash.to_vec(),
+                operation: ProjectedNftOperation::NotInlineDatum,
+                for_how_long: None,
+            };
+        }
+        DatumOption::Data(datum) => datum.0.encode_fragment().unwrap(),
+    };
+
+    let parsed = match cml_chain::plutus::PlutusData::from_bytes(datum.clone()) {
+        Ok(parsed) => parsed,
+        Err(_) => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
+                previous_utxo_tx_output_index: None,
+                address: vec![],
+                plutus_data: datum,
+                operation: ProjectedNftOperation::ParseError,
+                for_how_long: None,
+            }
+        }
+    };
+
+    let parsed = match projected_nft_sdk::State::try_from(parsed) {
+        Ok(parsed) => parsed,
+        Err(_) => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
+                previous_utxo_tx_output_index: None,
+                address: vec![],
+                plutus_data: datum,
+                operation: ProjectedNftOperation::ParseError,
+                for_how_long: None,
+            }
```

Comment:
Is it really okay to return this in the error state? When does the error state actually happen (when people create a Projected NFT datum that actually contains garbage and wasn't generated with our library?)

Reply:
yes